torchqdynamics/odeint.py

```diff
@@ -1,80 +1,122 @@
-import warnings
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import List, Optional
 
 import torch
 
 from .solver import AdaptativeStep, FixedStep
 
 
-def odeint(qsolver, y0, tsave, sensitivity='autograd', parameters=None):
-    # check arguments
-    tsave = torch.tensor(tsave)
-    check_tsave(tsave)
-    if (parameters is not None) and (sensitivity in [None, 'autograd']):
-        warnings.warn('Argument `parameters` was supplied in `odeint` but not used.')
+class ForwardQSolver(ABC):
+    @abstractmethod
+    def forward(self, t, dt, rho):
+        pass
 
-    # dispatch to appropriate odeint subroutine
-    if sensitivity is None:
-        return odeint_inplace(qsolver, y0, tsave)
-    elif sensitivity == 'autograd':
-        return odeint_main(qsolver, y0, tsave)
-    elif sensitivity == 'adjoint':
-        return odeint_adjoint(qsolver, y0, tsave, parameters)
 
+class GradientAlgorithm(Enum):
+    AUTOGRAD = 'autograd'  # gradient computed by torch
+    NONE = 'none'  # don't compute the gradients
+    ADJOINT = 'adjoint'  # compute the gradient using the adjoint method
+
+
+def odeint(
+    qsolver,
+    y0,
+    t_save: torch.Tensor,
+    exp_ops: Optional[List[torch.Tensor]] = None,
+    save_states: bool = True,
+    gradient_algorithm=GradientAlgorithm.AUTOGRAD,
+):
+    # check arguments
+    check_t_save(t_save)
 
-def odeint_main(qsolver, y0, tsave):
+    # dispatch to appropriate odeint subroutine
+    params = (qsolver, y0, t_save, exp_ops, save_states)
```

Comment:
Can we keep the terminology `parameters` for the model parameters to compute the gradients on? Maybe replace this with `args`?